pybart/new_matcher.py

```diff
@@ -0,0 +1,313 @@
+        for match_id, start, end in matches:
+            token_name = self.vocab.strings[match_id]
+            # assert that we matched no more than single token
+            if end - 1 != start:
+                continue
+            token = sentence.doc[start]
+            matched_tokens[token_name].append(token.id)
+
+        # extra token matching out of spacy's scope
+        matched_tokens = self._post_spacy_matcher(matched_tokens, sentence)
+
+        # reverse validate the 'optional' constraint
+        if len(self.required_tokens.difference(set(matched_tokens.keys()))) != 0:
+            raise RequiredTokenMismatch
+        
+        return matched_tokens
+
+
+class Matchers(NamedTuple):
+    token_matcher: TokenMatcher
+    global_matcher: GlobalMatcher
+
+
+class Match:
+    def __init__(self, matchers: Mapping[str, Matchers], sentence):
+        self.matchers = matchers
+        self.sentence = sentence
+    
+    def names(self) -> List[str]:
+        # return constraint-name list
+        return list(self.matchers.keys())
+    
+    def matches_for(self, name: str) -> Generator[MatchingResult, None, None]:
+        try:
+            # token match
+            matches = self.matchers[name].token_matcher.apply(self.sentence)
+        except Mismatch:
+            return
+        
+        # filter
+        yield from self.matchers[name].global_matcher.apply(matches, self.sentence)
+
+
+class NamedConstraint(NamedTuple):
+    name: str
+    constraint: Full
+
+
+# add TokenConstraints based on the other non-token constraints (optimization step).
+def preprocess_constraint(constraint: Full) -> Full:
+    # for each edge store the labels that could be filtered as incoming or outgoing token constraints
+    #   in the parent or child accordingly
+    outs = defaultdict(list)
+    ins = defaultdict(list)
+    for edge in constraint.edges:
+        # skip HasNoLabel as they check for non existing label between two nodes,
+        #   and if we add a token constraint it would be to harsh
+        if isinstance(edge.label, HasNoLabel):
+            continue
+        outs[edge.child].extend(list(edge.label))
+        ins[edge.parent].extend(list(edge.label))
+
+    # for each concat store the single words of the concat
+    #   with their correspondent token for token level WORD constraint
+    words = defaultdict(list)
+    for concat in constraint.concats:
+        zipped_concat = list(zip(*[tuple(t.split("_")) for t in concat.tuple_set]))
+        if isinstance(concat, TokenPair) or isinstance(concat, TokenTriplet):
+            words[concat.token1].append(zipped_concat[0])
+            words[concat.token2].append(zipped_concat[1])
+        if isinstance(concat, TokenTriplet):
+            words[concat.token3].append(zipped_concat[2])
+
+    # rebuild the constraint (as it is immutable)
+    tokens = []
+    for token in constraint.tokens:
+        incoming_edges = list(token.incoming_edges) + ins.get(token.id, [])
```

Comment:
should these be sets? or is it lists on purpose?

Reply:
@yoavg, these are lists on purpose as they represent the sequence of Label-constraints (and label strings or so) that is assigned to this token. So having set on constraint sequence is a bit weird I think. does it make sense?